File: python_solutions/chapter_08_recursion_and_dynamic_programming/problem_08_05_recursive_multiply.py
def recursive_multiply(a, b):
    A = max(a, b)
    B = min(a, b)
    return helper_recursive_multiply(A, B)


def helper_recursive_multiply(a, b):
    if b == 1:
        return a
    if b == 0:
        return 0
    half_b = int(b >> 1)
    if b > half_b + half_b:  # odd b. shifting with truncation eliminates a 1
        return a + helper_recursive_multiply(a << 1, half_b)  # odd case
    else:
        return helper_recursive_multiply(a << 1, half_b)  # even case

File: cpp_solutions/chapter_08_recursion_and_dynamic_programming/problem_08_05_recursiveMultiply.cpp
#include "problem_08_05_recursiveMultiply.h"
#include <algorithm>

namespace chapter_08{
    int recursiveMultiplyHelper(int a, int b){
        if(b == 1){
            return a;
        } else if (b <= 0){
            return 0;
        }
        int half_b = b >> 1;
        if(b > (half_b + half_b)){  // odd b. a 1 was removed during truncation
            return a + recursiveMultiplyHelper(a << 1, half_b);  // odd case
        } else {
            return recursiveMultiplyHelper(a << 1, half_b);  // even case
        }
    }

    int recursiveMultiply(int a, int b){
        int A = std::max(a, b);
        int B = std::min(a, b);
        return recursiveMultiplyHelper(A, B);
    }
}

